optaplanner-core/src/main/java/org/optaplanner/core/api/domain/common/DomainAccessType.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.optaplanner.core.api.domain.common;
+
+/**
+ * Determines how members (fields and methods) of
+ * the domain are accessed.
+ */
+public enum DomainAccessType {
+    /**
+     * Use reflection to read the member. Always work, but slow.
+     */
+    REFLECTION,
+
+    /**
+     * Uses Gizmo generated bytecode to access members (field and method)
+     * of the domain for performance, instead of reflection. When used without Quarkus,
+     * the planning annotations must be on public members and Gizmo must be added
+     * on the classpath or modulepath.
+     */
+    GIZMO
+}
```

Comment:
So, I was thinking... why does the user need to make this choice?

- We can detect if Gizmo is on the classpath/modulepath. In that case, we can just enable it.
- Likewise, we can detect if the user has annotations on non-public members.

Therefore - why not simply use Gizmo when present and possible, and use reflection for the rest? 

I'm sure I'm missing some consideration, but from where I'm standing, it seems like we could just make the best choice for the user, instead of forcing them to think about this.

Reply:
The main issue is unexpected performance changes; changing a member from public to private or vice versa could cause quite a change to performance and the user wouldn't understand why. Both of the above things are checkable and can be done on a per-member basis (which mean you could have a mix of Gizmo member accessors and Reflection member accessors). Removing the choice would fix the one remaining Reflection instance done when GIZMO is used (in the PlanningIdComparator, which is used in ConstraintMatch).

Reply:
Lukas, yes for predictability. There might be some other reasons.
The logic we came up with Radovan during the meeting is in this document (although a cat won't find its kittens in there):
  https://docs.google.com/document/d/1igSwaVuZA5_5ht-3nHmND6GBNvUQpDwvIoQMzshB7vg/edit

Reply:
Basically, if they explicitly ask for GIZMO or REFLECTION, that is what they get.
Otherwise, they get gizmo on quarkus and reflection otherwise. Simple & predictable.